File: gdrive_cli/paths.py
from __future__ import annotations

import os
from pathlib import Path


def config_home() -> Path:
    override = os.environ.get("XDG_CONFIG_HOME")
    base = Path(override).expanduser() if override else Path.home() / ".config"
    return base / "gdrive"


def data_home() -> Path:
    override = os.environ.get("XDG_DATA_HOME")
    base = Path(override).expanduser() if override else Path.home() / ".local" / "share"
    return base / "gdrive"


def config_file() -> Path:
    return config_home() / "config.json"


def legacy_token_file() -> Path:
    return data_home() / "token.json"


def token_dir() -> Path:
    return data_home() / "tokens"


def token_file_for_email(email: str) -> Path:
    return token_dir() / f"{email.strip().lower()}.json"


def state_dir() -> Path:
    return data_home() / "state"


def ensure_dirs() -> None:
    config_home().mkdir(parents=True, exist_ok=True, mode=0o700)
    data_home().mkdir(parents=True, exist_ok=True, mode=0o700)
    token_dir().mkdir(parents=True, exist_ok=True, mode=0o700)
    state_dir().mkdir(parents=True, exist_ok=True, mode=0o700)

File: internal/paths/paths.go
package paths

import (
	"os"
	"path/filepath"
	"strings"
)

func ConfigHome() string {
	base := os.Getenv("XDG_CONFIG_HOME")
	if base == "" {
		base = filepath.Join(homeDir(), ".config")
	}
	return filepath.Join(expandHome(base), "gdrive")
}

func DataHome() string {
	base := os.Getenv("XDG_DATA_HOME")
	if base == "" {
		base = filepath.Join(homeDir(), ".local", "share")
	}
	return filepath.Join(expandHome(base), "gdrive")
}

func ConfigFile() string {
	if override := os.Getenv("GDRIVE_CONFIG"); override != "" {
		return expandHome(override)
	}
	return filepath.Join(ConfigHome(), "config.json")
}

func TokenDir() string {
	return filepath.Join(DataHome(), "tokens")
}

func TokenFileForEmail(email string) string {
	return filepath.Join(TokenDir(), strings.ToLower(strings.TrimSpace(email))+".json")
}

func StateDir() string {
	return filepath.Join(DataHome(), "state")
}

func EnsureDirs() error {
	for _, dir := range []string{ConfigHome(), DataHome(), TokenDir(), StateDir()} {
		if err := os.MkdirAll(dir, 0o700); err != nil {
			return err
		}
	}
	return nil
}

func ExpandHome(path string) string {
	return expandHome(path)
}

func expandHome(path string) string {
	if path == "~" {
		return homeDir()
	}
	if strings.HasPrefix(path, "~/") {
		return filepath.Join(homeDir(), strings.TrimPrefix(path, "~/"))
	}
	return path
}

func homeDir() string {
	if home := os.Getenv("HOME"); home != "" {
		return home
	}
	if home, err := os.UserHomeDir(); err == nil {
		return home
	}
	return "."
}
